DIPs/DIP1005.md

```diff
@@ -0,0 +1,781 @@
+
+* Inside any function, all uses of `with` are statements and obey the current language rules.
+*  Everywhere else, `with (expression)` is not allowed. `with (Type)` and `with (TemplateInstance)` are always declarations and do not introduce a new scope. Lookup of symbols inside the `with` declarations is similar to lookup inside the `with` statement: symbols in the scopes of `Type` or `TemplateInstance` have priority (hide) symbols outside the `with` declaration.
+
+In addition, we propose the statement and declaration `with (import ImportList)`. `ImportList` is any syntactical construct currently accepted by the `import` declaration. The `with (import ImportList)` declaration obeys the following rules:
+
+* Inside any function, `with (Import ImportList)` is a statement that introduces a scope. Inside the `with`, lookup considers the `import` local to the declaration (similar to the current handling of nested `import`s).
+*  Everywhere else, `with (Import ImportList)` is always a declaration and does not introduce a new scope. Lookup of symbols is the same as for the statement case.
+
+This extension removes an unforced limitation of the current `with` syntax (allows it to occur at top level) and introduces a natural extension from symbol tables present in a type or template instance, to symbol tables imported from a module. The drawback of this choice is the potentially confusing handling of scopes: the `with` statement introduces a scope, whereas the `with` declaration does not.
+
+### The `with` Declaration
+
+The usual grammar of the `import ImportList;` declaration applies inside the `with (import ImportList)` declaration, with the following consequences:
+* The usual lookup rules apply, for example either `with (import std.range)` or the more precise `with (import std.range.primitives)` may be used to look up `isInputRange`.
+* Specific imports can be present as in `with (import std.range : isInputRange)` or `with (import std.range.primitives : isInputRange)`.
+* Renamed imports may be present as in `with (import std.range : isInput = isInputRange)`. This specification precludes the use of `isInputRange` and requires the use of `isInput` instead.
+
+The `static import` feature is also available with this syntax: `with (static import ImportList)`.
+
+Inline imports apply to all declarations (template or not) and may guard multiple declarations:
+
+```d
+with (import module_a : A, B)
+{
+    struct Widget(T = A) { ... }
+    alias C = B;
+}
+Widget!int g_widget;
+```
+
+As mentioned, `with` declarations do not introduce a scope so `Widget` above is visible outside the `with` declaration, but `A` and `B` are not.
+
+Inline imports apply to all declarations. This includes the `with` declaration itself, having the consequence that multiple `with` import declarations may be applied in a cascading manner:
+
+```d
+with (import module_a : A)
+with (import module_b : B)
+A fun(B) { ... }
```

Comment:
Can we please require the with import declarations to go after the function name and arguments, like template constraints?  It's bad enough that the function name can be preceded by an arbitrarily long return type and attributes, fishing it out of this sandwich is unnecessarily difficult.